decisions/2024-12-03_preferences.md

```diff
@@ -0,0 +1,148 @@
+
+The tables have a troubled relationship with sync. On initialisation, there is a lot of cruft such as all the central server datafile preferences being synced e.g. the default printer for the central server machine, various site version metadata records... Unlike most tables, there is no trigger code handling sync. It is handled by **explicit** calls to queue the pref record for sites whenever the pref record are modified and it is necessary to sync them, which is easy to get wrong or neglect to do. They do not go into the central change log.
+
+A big downside with these tables is the lack of strict typing around the shape of their pref data. It is often very hard to know what prefs exist in the system, why they exist and what shape their data is by looking at the code. It is often easier to analyse datafiles, dump prefs to JSON and text search to try find a relevant pref record if you're not sure what pref you're looking for. There is no enum or catalogue, and little to no dev documentation.
+
+#### OMS
+
+We currently just have a basic pref setup for store prefs. There is a `store_preference` table that has a column for each preference. This provides typing. There are some column limits of 1600 for postgres and ~32000 for sqlite, we're unlikely to hit those though.
+
+This doesn't cover things such as machine or user preferences by it self. To follow this pattern for other requirements, you'd have to make the tables `user_preference` and `machine_preference` and so on. If a future requirement, cross overs of those prefs might not really work without getting crazy e.g. a user pref for a particular store.
+
+## Options
+
+### Option 1 - Take the lessons from OG `pref` table by doing the same
+
+The OG pref table is battle tested for over a decade:
+
+- We create a `preference` (or `pref`?) table with `id`, `item`, `store_id`, `user_id`, `machine_id` and `data` fields.
+- The lack of typing that ailed OG can readily be addressed with with structs in OMS. These structs may carry through to the frontend gql types as appropriate.
+- Sync needs deeper thought, but at a glance: store prefs are easily identified and synced by the `store_id`. `user_id` could be added to `change_log` to handle that (if needed to sync... or we sync users and their prefs everywhere 😉). Perhaps it's OK to not sync and lose machine_id records.
+- We store the pref data in an JSON text field. We deserialize this with serde into structs for each pref as expected.
+  - This does mean that sync does not validate the `data` field, that validation occurs when the record is accessed. Though, we could quickly deserialize at sync buffer integration time just to check!
+- We can drop the `store_preference` table, but keep the repo the same for now, just querying from `preference` instead.
+
```

Comment:
```suggestion
- The list of preferences is defined by a single struct in the backend. The backend is responsible for merging Global, Store, and potentially if required `User Prefs/Machine Prefs`.
The single struct can have additional complexity, e.g. struct within structs as required. This may require additional mapping to graphql though.

```